lib/elixir/lib/registry.ex

```diff
@@ -587,14 +587,37 @@ defmodule Registry do
   def keys(registry, pid) when is_atom(registry) and is_pid(pid) do
     {kind, partitions, _, pid_ets, _} = info!(registry)
     {_, pid_ets} = pid_ets || pid_ets!(registry, pid, partitions)
-    keys = safe_lookup_second(pid_ets, pid)
+
+    keys = try do
+      spec = [{{pid, :'$1', :'$2'}, [], [{{:'$1', :'$2'}}]}]
+      :ets.select(pid_ets, spec)
+    catch
+      :error, :badarg -> []
+    end
+
+    # Handle the possibility of fake keys
+    keys = gather_keys(keys)
 
     cond do
       kind == :unique -> Enum.uniq(keys)
       true -> keys
     end
   end
 
+  defp gather_keys(keys, acc \\ [], fake \\ false)
```

Comment:
Please remove the default clause and call it with `gather_keys(keys, [], false)`. We rarely use default arguments for private functions in Elixir and we do so only if we have multiple callsites.